File: ios/Classes/MockDeviceKitHandler.swift
import Flutter

#if DEBUG
import MWDATMockDevice
#endif

class MockDeviceKitHandler {
    #if DEBUG
    private var pairedDevices: [String: MockDevice] = [:]
    #endif

    func pairRaybanMeta(result: @escaping FlutterResult) {
        #if DEBUG
        let device = MockDeviceKit.shared.pairRaybanMeta()
        let deviceId = device.deviceIdentifier
        pairedDevices[deviceId] = device
        result(["deviceId": deviceId])
        #else
        result(FlutterError(code: "MOCK_UNAVAILABLE", message: "MockDeviceKit is only available in debug builds", details: nil))
        #endif
    }

    func unpairDevice(deviceId: String, result: @escaping FlutterResult) {
        #if DEBUG
        guard let device = pairedDevices[deviceId] else {
            result(FlutterError(code: "MOCK_ERROR", message: "Device not found: \(deviceId)", details: nil))
            return
        }
        MockDeviceKit.shared.unpairDevice(device)
        pairedDevices.removeValue(forKey: deviceId)
        result(nil)
        #else
        result(FlutterError(code: "MOCK_UNAVAILABLE", message: "MockDeviceKit is only available in debug builds", details: nil))
        #endif
    }

    func deviceAction(deviceId: String, action: String, result: @escaping FlutterResult) {
        #if DEBUG
        guard let device = pairedDevices[deviceId] else {
            result(FlutterError(code: "MOCK_ERROR", message: "Device not found: \(deviceId)", details: nil))
            return
        }
        switch action {
        case "powerOn":
            device.powerOn()
        case "powerOff":
            device.powerOff()
        case "don":
            device.don()
        case "doff":
            device.doff()
        case "fold":
            if let glasses = device as? MockDisplaylessGlasses {
                glasses.fold()
            }
        case "unfold":
            if let glasses = device as? MockDisplaylessGlasses {
                glasses.unfold()
            }
        default:
            result(FlutterError(code: "MOCK_ERROR", message: "Unknown action: \(action)", details: nil))
            return
        }
        result(nil)
        #else
        result(FlutterError(code: "MOCK_UNAVAILABLE", message: "MockDeviceKit is only available in debug builds", details: nil))
        #endif
    }

    func setCameraFeed(deviceId: String, filePath: String, result: @escaping FlutterResult) {
        #if DEBUG
        guard let device = pairedDevices[deviceId] as? MockDisplaylessGlasses else {
            result(FlutterError(code: "MOCK_ERROR", message: "Device not found or not displayless glasses: \(deviceId)", details: nil))
            return
        }
        guard let cameraKit = device.getCameraKit() else {
            result(FlutterError(code: "MOCK_ERROR", message: "Camera kit not available", details: nil))
            return
        }
        let url = URL(fileURLWithPath: filePath)
        Task {
            await cameraKit.setCameraFeed(fileURL: url)
            await MainActor.run {
                result(nil)
            }
        }
        #else
        result(FlutterError(code: "MOCK_UNAVAILABLE", message: "MockDeviceKit is only available in debug builds", details: nil))
        #endif
    }

    func setCapturedImage(deviceId: String, filePath: String, result: @escaping FlutterResult) {
        #if DEBUG
        guard let device = pairedDevices[deviceId] as? MockDisplaylessGlasses else {
            result(FlutterError(code: "MOCK_ERROR", message: "Device not found or not displayless glasses: \(deviceId)", details: nil))
            return
        }
        guard let cameraKit = device.getCameraKit() else {
            result(FlutterError(code: "MOCK_ERROR", message: "Camera kit not available", details: nil))
            return
        }
        let url = URL(fileURLWithPath: filePath)
        Task {
            await cameraKit.setCapturedImage(fileURL: url)
            await MainActor.run {
                result(nil)
            }
        }
        #else
        result(FlutterError(code: "MOCK_UNAVAILABLE", message: "MockDeviceKit is only available in debug builds", details: nil))
        #endif
    }

    func dispose() {
        #if DEBUG
        pairedDevices.removeAll()
        #endif
    }
}

File: android/src/main/kotlin/com/meta/wearable/flutter/MockDeviceKitHandler.kt
package com.meta.wearable.flutter

import android.app.Application
import android.net.Uri
import com.meta.wearable.dat.mockdevice.MockDeviceKit
import com.meta.wearable.dat.mockdevice.api.MockDeviceKitInterface
import com.meta.wearable.dat.mockdevice.api.MockRaybanMeta
import io.flutter.plugin.common.MethodChannel

class MockDeviceKitHandler(private val application: Application) {
    private val mockDeviceKit: MockDeviceKitInterface by lazy {
        MockDeviceKit.getInstance(application.applicationContext)
    }
    private val pairedDevices = mutableMapOf<String, MockRaybanMeta>()

    fun pairRaybanMeta(result: MethodChannel.Result) {
        try {
            val mockDevice = mockDeviceKit.pairRaybanMeta()
            val deviceId = System.currentTimeMillis().toString()
            pairedDevices[deviceId] = mockDevice
            result.success(mapOf("deviceId" to deviceId))
        } catch (e: Exception) {
            result.error("MOCK_ERROR", "Failed to pair mock device: ${e.message}", null)
        }
    }

    fun unpairDevice(deviceId: String, result: MethodChannel.Result) {
        try {
            val device = pairedDevices[deviceId]
            if (device != null) {
                mockDeviceKit.unpairDevice(device)
                pairedDevices.remove(deviceId)
                result.success(null)
            } else {
                result.error("MOCK_ERROR", "Device not found: $deviceId", null)
            }
        } catch (e: Exception) {
            result.error("MOCK_ERROR", "Failed to unpair device: ${e.message}", null)
        }
    }

    fun deviceAction(deviceId: String, action: String, result: MethodChannel.Result) {
        try {
            val device = pairedDevices[deviceId]
            if (device == null) {
                result.error("MOCK_ERROR", "Device not found: $deviceId", null)
                return
            }
            when (action) {
                "powerOn" -> device.powerOn()
                "powerOff" -> device.powerOff()
                "don" -> device.don()
                "doff" -> device.doff()
                "fold" -> device.fold()
                "unfold" -> device.unfold()
                else -> {
                    result.error("MOCK_ERROR", "Unknown action: $action", null)
                    return
                }
            }
            result.success(null)
        } catch (e: Exception) {
            result.error("MOCK_ERROR", "Failed to execute action: ${e.message}", null)
        }
    }

    fun setCameraFeed(deviceId: String, filePath: String, result: MethodChannel.Result) {
        try {
            val device = pairedDevices[deviceId]
            if (device == null) {
                result.error("MOCK_ERROR", "Device not found: $deviceId", null)
                return
            }
            device.getCameraKit().setCameraFeed(Uri.parse(filePath))
            result.success(null)
        } catch (e: Exception) {
            result.error("MOCK_ERROR", "Failed to set camera feed: ${e.message}", null)
        }
    }

    fun setCapturedImage(deviceId: String, filePath: String, result: MethodChannel.Result) {
        try {
            val device = pairedDevices[deviceId]
            if (device == null) {
                result.error("MOCK_ERROR", "Device not found: $deviceId", null)
                return
            }
            device.getCameraKit().setCapturedImage(Uri.parse(filePath))
            result.success(null)
        } catch (e: Exception) {
            result.error("MOCK_ERROR", "Failed to set captured image: ${e.message}", null)
        }
    }

    fun dispose() {
        pairedDevices.clear()
    }
}
